Validate that path has valid extension when saving a screenshot (closes #400)

abotalov opened the pull request:
Webdriver spec [says that only PNG is supported](https://w3c.github.io/webdriver/webdriver-spec.html#dfn-encode-as-base64-a-canvas-element).

abotalov committed:
Validate that path has valid extension when saving a screenshot (closes #400)

p0deje commented:
Looks good to me, however I see that there is a failing spec now - https://ci.appveyor.com/project/p0deje/watir/build/%23175/job/nx6srh2qh8csua9v, can you fix it?

abotalov committed:
Use SecureRandom to generate a filename in screenshot specs.
Time.now.to_i returns duplicate numbers if invoked more than once
at the same second.

titusfortner commented:
Except Watir isn't claiming to be W3C compliant right now. jpg works just fine in chromedriver & geckodriver, and I don't think this version of Watir should prevent users from that format if they want.

p0deje commented:
#400 says that even if you specify .jpg it will still be PNG image when used with chromedriver.

abotalov commented:
@titusfortner Relevant commands in bridges don't take any format parameters, see https://github.com/SeleniumHQ/selenium/blob/ceaf3da79542024becdda5953059dfbb96fb3a90/rb/lib/selenium/webdriver/common/driver_extensions/takes_screenshot.rb#L46
.NET has an ability to save screenshot as jpg by converting png to jpg - https://github.com/SeleniumHQ/selenium/blob/1e221937594c8ac4895899024927575aa6015a60/dotnet/src/webdriver/Screenshot.cs

abotalov committed:
Warn about non-PNG extensions instead of raising an error because of Semver

abotalov commented:
I have now changed code to warn about invalid extension rather than raising an error because of Semver.

titusfortner commented:
Oh, interesting. Shouldn't this verification happen in Selenium, then?

titusfortner commented:
Also, defining a constant array with one value seems unnecessary to me.

abotalov commented:
Defining a constant has the aim to allow users to add additional extensions if they would like to save PNG screenshots with different extensions.

p0deje commented:
We should not allow mutating constants, we should instead freeze it.

titusfortner commented:
That's what I'm saying, we shouldn't have mutable constants, as if it isn't mutable then it is overkill for one item. We could create a method that could be subclassed, or create a keyword to override exception, or we leave it as a warning.

I think the method itself should just set it to png & PNG. Or maybe we go the other way and complain if it is jpg, jpeg, etc, since that's probably the real concern.

As for Watir vs selenium, Watir is just a pass through here. If selenium users think they are getting a jpg, but aren't, they should at least be warned.

abotalov commented:
This is now handled in Selenium - https://github.com/SeleniumHQ/selenium/issues/4108

titusfortner commented:
Fixed in ruby or just python?

abotalov commented:
@titusfortner I meant "It's planned to be handled in Selenium" :)
